ngraph/core/include/openvino/core/shared_any.hpp

```diff
@@ -0,0 +1,337 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+#include <ostream>
+#include <typeinfo>
+
+#include "openvino/core/core_visibility.hpp"
+#include "openvino/core/except.hpp"
+#include "openvino/core/type.hpp"
+
+namespace ov {
+
+/**
+ * @brief This class represents API of internal storage that can be returned from DLL
+ * using std::shared_ptr
+ *
+ */
+struct OPENVINO_API SharedAnyImpl : std::enable_shared_from_this<SharedAnyImpl> {
+    using Ptr = std::shared_ptr<SharedAnyImpl>;
+    virtual const std::type_info& get_type_info() const = 0;
+    virtual void print(std::ostream& os) const = 0;
+    virtual void* get() = 0;
+    const void* get() const {
+        return const_cast<SharedAnyImpl*>(this)->get();
+    }
+    virtual bool equal(const SharedAnyImpl& other) const = 0;
+
+    /**
+     * @brief Checks the type of value
+     * @tparam T Type of value
+     * @return true if type of value is correct
+     */
+    template <typename T>
+    bool is() const {
+        return get_type_info() == typeid(T);
+    }
+
+    /**
+     * @brief Cast to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <typename T>
+    T& as() & {
+        OPENVINO_ASSERT(is<T>(), "Bad cast from", get_type_info().name(), " to ", typeid(T).name());
+        return *static_cast<typename std::decay<T>::type*>(get());
+    }
+
+    /**
+     * @brief Cast to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <typename T>
+    const T& as() const& {
+        OPENVINO_ASSERT(is<T>(), "Bad cast from", get_type_info().name(), " to ", typeid(T).name());
+        return *static_cast<typename std::decay<T>::type*>(get());
+    }
+
+    /**
+     * @brief Cast to specified type
+     * @tparam T type
+     * @return casted object
+     */
+    template <typename T>
+    T&& as() && {
+        OPENVINO_ASSERT(is<T>(), "Bad cast from", get_type_info().name(), " to ", typeid(T).name());
+        return std::move(*static_cast<typename std::decay<T>::type*>(get()));
+    }
+
+protected:
+    ~SharedAnyImpl() = default;
+};
+
+/**
+ * @brief This class represents generic shared value object
+ */
+class SharedAny {
+    template <class T, class EqualTo>
+    struct CheckOperatorEqual {
```

Comment:
VariantWrappers have some methods for merge different object.
Am I right that we don't support such functionality here?

CC @GlebKazantaev